### PR TITLE
rgw_sal_motr : [CORTX-30347]: MotrMultipartUpload: Add object tags with CreateMultipartUpload API

### PR DESCRIPTION
**Pull request by suryakumar1024**
Problem:
User with valid permission can create tag set while creating multipart upload, Which is not happening. Tagging params are not reflected back in the get object tag details.

Fix Posted:
Fixed by adding tagging key and value to attrs variable in rgw_sal_motr layer init() method.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by suryakumar1024** [comment]
1N K8 setup deployment success

CORTX Cluster Setup SUCCESS
Please check [cluster setup logs](https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/setup-cortx-rgw-cluster/6143//console) for more info

Cluster Status:
==================== Image Details ====================

ghcr.io/seagate/consul:1.11.4
cortx-docker.colo.seagate.com/seagate/cortx-control:2.0.0-6522-custom-ci
cortx-docker.colo.seagate.com/seagate/cortx-data:2.0.0-6522-custom-ci
cortx-docker.colo.seagate.com/seagate/cortx-control:2.0.0-6522-custom-ci
ghcr.io/seagate/kafka:3.0.0-debian-10-r97
cortx-docker.colo.seagate.com/seagate/cortx-rgw:2.0.0-6522-custom-ci
ghcr.io/seagate/zookeeper:3.8.0-debian-10-r9

==================== POD Status ====================

NAME                                                 READY   STATUS    RESTARTS   AGE     IP               NODE                                    NOMINATED NODE   READINESS GATES
cortx-consul-client-mp8sf                            1/1     Running   0          3m53s   172.18.221.118   ssc-vm-g2-rhev4-3137.colo.seagate.com              
cortx-consul-server-0                                1/1     Running   0          3m58s   172.18.221.106   ssc-vm-g2-rhev4-3137.colo.seagate.com              
cortx-control-5cd77f4bc4-5bzm4                       1/1     Running   0          3m8s    172.18.221.120   ssc-vm-g2-rhev4-3137.colo.seagate.com              
cortx-data-ssc-vm-g2-rhev4-3137-5c69dddbd6-w5xks     4/4     Running   0          2m34s   172.18.221.117   ssc-vm-g2-rhev4-3137.colo.seagate.com              
cortx-ha-57b46c5dfc-bhr48                            3/3     Running   0          21s     172.18.221.77    ssc-vm-g2-rhev4-3137.colo.seagate.com              
cortx-kafka-0                                        1/1     Running   0          4m47s   172.18.221.108   ssc-vm-g2-rhev4-3137.colo.seagate.com              
cortx-server-ssc-vm-g2-rhev4-3137-54d784d879-289n8   2/2     Running   0          90s     172.18.221.70    ssc-vm-g2-rhev4-3137.colo.seagate.com              
cortx-zookeeper-0                                    1/1     Running   0          4m47s   172.18.221.104   ssc-vm-g2-rhev4-3137.colo.seagate.com              

--------------- Sleeping for 1min before checking hctl status.... ---------------


--------------- hctl status ---------------

Mon Jun  6 11:16:46 MDT 2022
Bytecount:
    critical : 0
    damaged : 0
    degraded : 0
    healthy : 0
Data pool:
    # fid name
    0x6f00000000000001:0x37 'storage-set-1__sns'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0x54 'Profile_the_pool': 'storage-set-1__sns' 'storage-set-1__dix' None
Services:
    cortx-server-headless-svc-ssc-vm-g2-rhev4-3137 
    [started]  hax                 0x7200000000000001:0x2d         inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-3137@22001
    [started]  rgw_s3              0x7200000000000001:0x30         inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-3137@21001
    cortx-data-headless-svc-ssc-vm-g2-rhev4-3137  (RC)
    [started]  hax                 0x7200000000000001:0x7          inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-3137@22001
    [started]  ioservice           0x7200000000000001:0xa          inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-3137@21001
    [started]  ioservice           0x7200000000000001:0x19         inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-3137@21002
    [started]  confd               0x7200000000000001:0x28         inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-3137@22002

==================== Time taken for service to start 0 mins ====================

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -4003,11 +4026,18 @@ int MotrMultipartUpload::complete(const DoutPrefixProvider *dpp,
...
+  rgw::sal::Attrs temp_attrs;
   rgw_bucket_dir_entry ent;
   bufferlist& blr = bl;
   auto ent_iter = blr.cbegin();
   ent.decode(ent_iter);
+  decode(temp_attrs, ent_iter);
 
+  // Add tag to attrs[RGW_ATTR_TAGS] key only if temp_attrs has tagging info
+  if (auto aiter = temp_attrs.find(RGW_ATTR_TAGS);
+      aiter != temp_attrs.end()) {
```
Comment (rahul-tripathi-git) [inline src/rgw/rgw_sal_motr.cc:4038]: if (temp_attrs.find(RGW_ATTR_TAGS) != temp_attrs.end())
or
if (temp_attrs.count(RGW_ATTR_TAGS)) // returns 0 if key isnot present
or
if (temp_attrs.contains(RGW_ATTR_TAGS)) // membership check

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -3691,7 +3694,27 @@ int MotrMultipartUpload::init(const DoutPrefixProvider *dpp, optional_yield y,
     ent.meta.mtime = ceph::real_clock::now();
     ent.meta.user_data.assign(mpbl.c_str(), mpbl.c_str() + mpbl.length());
     ent.encode(bl);
-
+    std::unique_ptr<RGWObjTags> obj_tags;
+    req_state *s = (req_state *) obj_ctx->get_private();
+    /* handle object tagging */
+    // Verify tags exists and add to attrs
+    auto tag_str_bool = s->info.env->exists("HTTP_X_AMZ_TAGGING");
+    if (tag_str_bool){
```
Comment (rahul-tripathi-git) [inline src/rgw/rgw_sal_motr.cc:3702]: if (s->info.env->exists("HTTP_X_AMZ_TAGGING"){

Comment (suryakumar1024) [reply src/rgw/rgw_sal_motr.cc]: Updated in recent patch

File: src/rgw/rgw_sal_motr.cc [hunk 3]
```diff
@@ -3691,7 +3694,26 @@ int MotrMultipartUpload::init(const DoutPrefixProvider *dpp, optional_yield y,
...
+    std::unique_ptr<RGWObjTags> obj_tags;
+    req_state *s = (req_state *) obj_ctx->get_private();
+    /* handle object tagging */
+    // Verify tags exists and add to attrs
+    if (s->info.env->exists("HTTP_X_AMZ_TAGGING")){
+      auto tag_str = s->info.env->get("HTTP_X_AMZ_TAGGING");
+      obj_tags = std::make_unique<RGWObjTags>();
+      int ret = obj_tags->set_from_string(tag_str);
+      if (ret < 0){
+        ldpp_dout(dpp,0) << "setting obj tags failed with rc=" << ret << dendl;
```
Comment (selvakumaar5496) [inline src/rgw/rgw_sal_motr.cc:3706]: space after comma

**Review by selvakumaar5496 — APPROVED** [review]

**Comment by sjain09** [comment]
"MotrMultipartUpload: " string looks unwanted in description. Please remove it

**Comment by sjain09** [comment]
There are too many commits for a small change. Are all commits relevant to this change?

**Comment by suryakumar1024** [comment]
> There are too many commits for a small change. Are all commits relevant to this change?

Yes, @sjain09 All commits are relevant

**Comment by suryakumar1024** [comment]
> "MotrMultipartUpload: " string looks unwanted in description. Please remove it

Updated description

**Review by sjain09 — APPROVED** [review]

**Review by cdeshmukh — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc (hunk 3 above)
Comment (suryakumar1024) [reply src/rgw/rgw_sal_motr.cc]: Updated in recent patch

**Comment by suryakumar1024** [comment]
Tested by deploying:

Custom-CI image
https://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/generic/job/custom-ci/6573/console

Docket images
cortx-docker.colo.seagate.com/seagate/cortx-rgw:2.0.0-6573-custom-ci
cortx-docker.colo.seagate.com/seagate/cortx-data:2.0.0-6573-custom-ci
cortx-docker.colo.seagate.com/seagate/cortx-all:2.0.0-6573-custom-ci
cortx-docker.colo.seagate.com/seagate/cortx-control:2.0.0-6573-custom-ci
image


Deployment success status
https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/setup-cortx-rgw-cluster/6380/
image

Cluster Status:
==================== Image Details ====================

ghcr.io/seagate/consul:1.11.4
cortx-docker.colo.seagate.com/seagate/cortx-control:2.0.0-6573-custom-ci
cortx-docker.colo.seagate.com/seagate/cortx-data:2.0.0-6573-custom-ci
cortx-docker.colo.seagate.com/seagate/cortx-control:2.0.0-6573-custom-ci
ghcr.io/seagate/kafka:3.0.0-debian-10-r97
cortx-docker.colo.seagate.com/seagate/cortx-rgw:2.0.0-6573-custom-ci
ghcr.io/seagate/zookeeper:3.8.0-debian-10-r9

==================== POD Status ====================

NAME                                              READY   STATUS    RESTARTS        AGE     IP               NODE                                 NOMINATED NODE   READINESS GATES
cortx-consul-client-5sn6b                         1/1     Running   0               6m39s   172.16.34.7      ssc-vm-rhev4-1730.colo.seagate.com              
cortx-consul-client-jg764                         1/1     Running   0               7m15s   172.16.120.142   ssc-vm-rhev4-1729.colo.seagate.com              
cortx-consul-client-pj6gk                         1/1     Running   0               6m29s   172.16.254.208   ssc-vm-rhev4-1728.colo.seagate.com              
cortx-consul-server-0                             1/1     Running   0               5m16s   172.16.120.139   ssc-vm-rhev4-1729.colo.seagate.com              
cortx-consul-server-1                             1/1     Running   0               6m26s   172.16.254.209   ssc-vm-rhev4-1728.colo.seagate.com              
cortx-consul-server-2                             1/1     Running   0               7m19s   172.16.34.6      ssc-vm-rhev4-1730.colo.seagate.com              
cortx-control-6459547ccb-pfmmk                    1/1     Running   0               4m13s   172.16.120.143   ssc-vm-rhev4-1729.colo.seagate.com              
cortx-data-ssc-vm-rhev4-1728-6c86cb586-cf47t      4/4     Running   0               3m29s   172.16.254.204   ssc-vm-rhev4-1728.colo.seagate.com              
cortx-data-ssc-vm-rhev4-1729-6fcc5954b5-9z6pm     4/4     Running   0               3m27s   172.16.120.145   ssc-vm-rhev4-1729.colo.seagate.com              
cortx-data-ssc-vm-rhev4-1730-bff8c668c-qtd9m      4/4     Running   0               3m26s   172.16.34.4      ssc-vm-rhev4-1730.colo.seagate.com              
cortx-ha-5f46f5599f-wzqpr                         3/3     Running   0               33s     172.16.254.202   ssc-vm-rhev4-1728.colo.seagate.com              
cortx-kafka-0                                     1/1     Running   1 (7m53s ago)   8m41s   172.16.254.198   ssc-vm-rhev4-1728.colo.seagate.com              
cortx-kafka-1                                     1/1     Running   0               8m41s   172.16.120.140   ssc-vm-rhev4-1729.colo.seagate.com              
cortx-kafka-2                                     1/1     Running   0               8m41s   172.16.34.63     ssc-vm-rhev4-1730.colo.seagate.com              
cortx-server-ssc-vm-rhev4-1728-556864c9bf-rhg28   2/2     Running   0               2m4s    172.16.254.211   ssc-vm-rhev4-1728.colo.seagate.com              
cortx-server-ssc-vm-rhev4-1729-b6cb8b7bc-gs782    2/2     Running   0               2m3s    172.16.120.135   ssc-vm-rhev4-1729.colo.seagate.com              
cortx-server-ssc-vm-rhev4-1730-8497596ccf-4lc5l   2/2     Running   0               2m2s    172.16.34.15     ssc-vm-rhev4-1730.colo.seagate.com              
cortx-zookeeper-0                                 1/1     Running   0               8m41s   172.16.254.205   ssc-vm-rhev4-1728.colo.seagate.com              
cortx-zookeeper-1                                 1/1     Running   0               8m41s   172.16.120.138   ssc-vm-rhev4-1729.colo.seagate.com              
cortx-zookeeper-2                                 1/1     Running   0               8m41s   172.16.34.5      ssc-vm-rhev4-1730.colo.seagate.com              

--------------- Sleeping for 1min before checking hctl status.... ---------------


--------------- hctl status ---------------

Wed Jun  8 11:11:55 MDT 2022
Bytecount:
    critical : 0
    damaged : 0
    degraded : 0
    healthy : 0
Data pool:
    # fid name
    0x6f00000000000001:0x9f 'storage-set-1__sns'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xde 'Profile_the_pool': 'storage-set-1__sns' 'storage-set-1__dix' None
Services:
    cortx-data-headless-svc-ssc-vm-rhev4-1729 
    [started]  hax                 0x7200000000000001:0x2f         inet:tcp:cortx-data-headless-svc-ssc-vm-rhev4-1729@22001
    [started]  ioservice           0x7200000000000001:0x32         inet:tcp:cortx-data-headless-svc-ssc-vm-rhev4-1729@21001
    [started]  ioservice           0x7200000000000001:0x41         inet:tcp:cortx-data-headless-svc-ssc-vm-rhev4-1729@21002
    [started]  confd               0x7200000000000001:0x50         inet:tcp:cortx-data-headless-svc-ssc-vm-rhev4-1729@22002
    cortx-data-headless-svc-ssc-vm-rhev4-1730  (RC)
    [started]  hax                 0x7200000000000001:0x7          inet:tcp:cortx-data-headless-svc-ssc-vm-rhev4-1730@22001
    [started]  ioservice           0x7200000000000001:0xa          inet:tcp:cortx-data-headless-svc-ssc-vm-rhev4-1730@21001
    [started]  ioservice           0x7200000000000001:0x19         inet:tcp:cortx-data-headless-svc-ssc-vm-rhev4-1730@21002
    [started]  confd               0x7200000000000001:0x28         inet:tcp:cortx-data-headless-svc-ssc-vm-rhev4-1730@22002
    cortx-data-headless-svc-ssc-vm-rhev4-1728 
    [started]  hax                 0x7200000000000001:0x57         inet:tcp:cortx-data-headless-svc-ssc-vm-rhev4-1728@22001
    [started]  ioservice           0x7200000000000001:0x5a         inet:tcp:cortx-data-headless-svc-ssc-vm-rhev4-1728@21001
    [started]  ioservice           0x7200000000000001:0x69         inet:tcp:cortx-data-headless-svc-ssc-vm-rhev4-1728@21002
    [started]  confd               0x7200000000000001:0x78         inet:tcp:cortx-data-headless-svc-ssc-vm-rhev4-1728@22002
    cortx-server-headless-svc-ssc-vm-rhev4-1730 
    [started]  hax                 0x7200000000000001:0x7d         inet:tcp:cortx-server-headless-svc-ssc-vm-rhev4-1730@22001
    [started]  rgw_s3              0x7200000000000001:0x80         inet:tcp:cortx-server-headless-svc-ssc-vm-rhev4-1730@21001
    cortx-server-headless-svc-ssc-vm-rhev4-1729 
    [started]  hax                 0x7200000000000001:0x85         inet:tcp:cortx-server-headless-svc-ssc-vm-rhev4-1729@22001
    [started]  rgw_s3              0x7200000000000001:0x88         inet:tcp:cortx-server-headless-svc-ssc-vm-rhev4-1729@21001
    cortx-server-headless-svc-ssc-vm-rhev4-1728 
    [started]  hax                 0x7200000000000001:0x8d         inet:tcp:cortx-server-headless-svc-ssc-vm-rhev4-1728@22001
    [started]  rgw_s3              0x7200000000000001:0x90         inet:tcp:cortx-server-headless-svc-ssc-vm-rhev4-1728@21001

==================== Time taken for service to start 0 mins ====================

**Pull request merged**